h2o-core/src/main/java/water/rapids/BinaryMerge.java

```diff
@@ -369,15 +412,146 @@ private boolean leftKeyEqual(byte x[][], long xi, long yi) {
+      long globalRowNumber = _leftKO.at8order(leftInd);
+      int chkIdx = _leftSB._vec.elem2ChunkIdx(globalRowNumber);// obtain chunk index of row of interest
+      _leftKO._perNodeNumRowsToFetch[_leftSB._chunkNode[chkIdx]]++;
+      if (_allLeft && binarySearchLen==0)
+        continue; // no need to fill out _ret1st and _retLen if there is no match in rite frame
+      long fLocationIndex = leftInd-(_leftFrom+1);
+      final int batchIndex = (int) fLocationIndex/_retBatchSize;
+      final int batchOffset = (int) fLocationIndex%_retBatchSize;
+      _ret1st[batchIndex][batchOffset] = rLow+1;  // 0 is to denote no match between left and rite frame
+      _retLen[batchIndex][batchOffset] = leftFrameIterate?binarySearchLen:iterLen;
+    }
+    // fill in rite frame and add rows to final merged frame if appropriate
+    if (!(_allLeft && binarySearchLen==0)) {
+      long riteLen = leftFrameIterate?binarySearchLen:(leftHigh - leftLow); // number of duplication in right frame
+      for (long riteInd = 0; riteInd < riteLen; riteInd++) {
+        long loc = rLow + riteInd;
+        long globalRowNumber = _riteKO.at8order(loc);
+        int chkIdx = _riteSB._vec.elem2ChunkIdx(globalRowNumber);
+        _riteKO._perNodeNumRowsToFetch[_riteSB._chunkNode[chkIdx]]++;
+      }
+    }
+  }
+
+  private boolean frameKeyEqual(byte x[][], long xi, long yi, KeyOrder fKO, FFSB fSB) {
+    byte xbatch[] = x[(int)(xi / fKO._batchSize)];
+    byte ybatch[] = x[(int)(yi / fKO._batchSize)];
+    int xoff = (int)(xi % fKO._batchSize) * fSB._keySize;
+    int yoff = (int)(yi % fKO._batchSize) * fSB._keySize;
+    int i=0;
+    while (i<fSB._keySize && xbatch[xoff++] == ybatch[yoff++]) i++;
+    return(i==fSB._keySize);
+  }
+  
+  private long[] bsearchRiteMatch(long iterIndex, long[] midSearchInd, boolean leftFIter) {
+    while (midSearchInd[1] < (midSearchInd[2]-1)) { // go through the appropriate range of rite frame to find match
+      midSearchInd[0] = midSearchInd[1]+(midSearchInd[2]-midSearchInd[1])/2;
+      int cmp = leftFIter?keycmp(_leftKO._key, iterIndex, _leftKO, _leftSB, _riteKO._key, midSearchInd[0], _riteKO, _riteSB):
+              keycmp(_riteKO._key, iterIndex, _riteKO, _riteSB, _leftKO._key, midSearchInd[0], _leftKO, _leftSB);// 0 equal, <0 smaller, >0 bigger
+      if (cmp < 0) { // iterate frame value lower than binarysearch value, shrink binarysearch frame index down to midRite
+        midSearchInd[2] = midSearchInd[0];
+      } else if (cmp > 0) { // iterate frame value higher than binarysearch value, move up binarysearch frame index
+        midSearchInd[1] = midSearchInd[0];
+      } else {  // iterate and binarysearch frame key match
+        return midSearchInd;
+      }      
+    }
+    midSearchInd[0] = -1;  // no match found
+    return midSearchInd;
+  }
+  
+  //*********  Not used, but DO NOT DELETE ***********
```

Comment:
why not? you clearly replaced it with binaryMergeOneMSB

Reply:
Honza:  They have done something clever here that I do not want to lose until I prove that their timing is worse than mine.  Please indulge me.

Reply:
Code removed.